contrib/opencensus-ext-azure/opencensus/ext/azure/metrics_exporter/__init__.py

```diff
@@ -12,31 +12,219 @@
+                    # identified by it's label values
+                    for time_series in metric.time_series:
+                        # Using stats, time_series should
+                        # only have one point which contains
+                        # the aggregated value
+                        data_point = self.create_data_points(
+                            time_series, md)[0]
+                        # The timestamp is when the metric was recorded
+                        time_stamp = time_series.points[0].timestamp
+                        # Get the properties using label keys from metric
+                        # and label values of the time series
+                        properties = self.create_properties(time_series, md)
+                        envelopes.append(self.create_envelope(data_point,
+                                                              time_stamp,
+                                                              properties))
+                        # Send data in batches of max_batch_size
+                        if len(envelopes) == self.max_batch_size:
+                            self._transmit_without_retry(envelopes)
+                            del envelopes[:]
+            # if leftover data points in envelopes, send them all
+            if envelopes:
+                self._transmit_without_retry(envelopes)
+
+    def create_data_points(self, time_series, metric_descriptor):
+        """Convert an metric's OC time series to list of Azure data points."""
+        data_points = []
+        for point in time_series.points:
+            data_point = DataPoint(ns=metric_descriptor.name,
+                                   name=metric_descriptor.name,
+                                   value=point.value.value)
+            data_points.append(data_point)
+        return data_points
+
+    def create_properties(self, time_series, metric_descriptor):
+        properties = {}
+        # We construct a properties map from the
+        # label keys and values
+        # We assume the ordering is already correct
+        for i in range(len(metric_descriptor.label_keys)):
+            if time_series.label_values[i].value is None:
+                value = "None"
+            else:
+                value = time_series.label_values[i].value
+            properties[metric_descriptor.label_keys[i].key] = value
+        return properties
+
+    def create_envelope(self, data_point, time_stamp, properties):
+        envelope = Envelope(
+            iKey=self.options.instrumentation_key,
+            tags=dict(utils.azure_monitor_context),
+            time=time_stamp.isoformat(),
+        )
+        envelope.name = "Microsoft.ApplicationInsights.Metric"
+        data = MetricData(
+            metrics=[data_point],
+            properties=properties
+        )
+        envelope.data = Data(baseData=data, baseType="MetricData")
+        return envelope
+
+    def _transmit_without_retry(self, envelopes):
```

Comment:
Leave a todo comment here.

Reply:
The TODO should cover:
1. consolidate with transport logic (instead of duplicating code)
2. handle failures properly

Reply:
Done.